src/emit/binary.rs

```diff
@@ -603,40 +594,15 @@ impl<'a> Binary<'a> {
         ns: &Namespace,
     ) -> FunctionType<'a> {
         // function parameters
-        let mut args = params
+        let args = params
             .iter()
             .map(|ty| self.llvm_var_ty(ty, ns).into())
             .collect::<Vec<BasicMetadataTypeEnum>>();
 
-        // add return values
-        for ty in returns {
-            args.push(if ty.is_reference_type(ns) && !ty.is_contract_storage() {
-                self.llvm_type(ty, ns)
-                    .ptr_type(AddressSpace::Generic)
-                    .ptr_type(AddressSpace::Generic)
-                    .into()
-            } else {
-                self.llvm_type(ty, ns)
-                    .ptr_type(AddressSpace::Generic)
-                    .into()
-            });
+        match returns.iter().next() {
+            Some(ret) => self.llvm_type(ret, ns).fn_type(&args, false),
+            None => self.context.void_type().fn_type(&args, false),
         }
-
-        // On Solana, we need to pass around the accounts
-        if ns.target == Target::Solana {
-            args.push(
-                self.module
-                    .get_struct_type("struct.SolParameters")
-                    .unwrap()
-                    .ptr_type(AddressSpace::Generic)
-                    .into(),
-            );
-        }
-
-        // Solana return type should be 64 bit, 32 bit on wasm
-        self.return_values[&ReturnCode::Success]
-            .get_type()
-            .fn_type(&args, false)
```

Comment:
Soroban functions always accept and return `i64` values, where those values are a Soroban RawVal, a 64-bit encoded value that Soroban contracts need to unpack. Soroban doesn't have a return code, and doesn't write return values to imported memory, so I commented this code out.

A RawVal is a 4-bit tag space with a 60-bit value space. Small values like 32-bit integers are encoded directly into the 64-bit value. Larger values like byte arrays, vectors, maps, are a 32-bit handle embedded into the RawVal, and there are host functions that can be used to extract elements of the larger value.

This pull request doesn't contain adding RawVal decode/encode for types, so when I was trying it out the uint64 values the contract were receiving were the RawVal encoded value, not the user value.

Reply:
>  I assume we'd inject those instructions into the beginning of functions, or as a bitcode as a form of stdlib.

For Borsh (Solana) and SCALE (Substrate / ink!), we have our own encoder and decoder implementations in `codegen`. Adding Soroban as a target would imply implementing an encoder for your encoding schema as well, which should solve the problem (encoding is needed in multiple places, e.g. events, calls, return values).

Reply:
Great, yes that seems like the best place to implement it. Thanks 👍🏻.